Fix updates of sleep_time in DistributedAsyncInsertDirectoryQueue

Blackmorse opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

When using background inserts into tables with the `Distributed` engine, delays caused by repeated errors when sending data to remote shards now decrease once the errors are resolved. Previously, the delay would only increase and never reset. 




<!-- ch-version-info:start -->
### Version info
- Merged into: `26.7.1.1139` (included in `26.7` and later)
<!-- ch-version-info:end -->

Blackmorse committed:
Fix updates of sleep_time in DistributedAsyncInsertDirectoryQueue

tuanpach commented on src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp:
```diff
@@ -226,19 +236,14 @@ void DistributedAsyncInsertDirectoryQueue::run()
                     status.error_count /= 2;
                     last_decrease_time = now;
+
+                    updateSleepTime();
```
> So, yes, if we processed all the files, then we break the loop and we sleep.

When `pending_files.isFinished()`, we won't reschedule the task.

```
if (!pending_files.isFinished() && do_sleep)
        task_handle->scheduleAfter(sleep_time.count());
```

> But if we are not experiencing errors anymore, then the sleep time will never updated, but it should

If we don't reschedule the task, updating `sleep_time` seems to have no effect.

Could you explain it more to me?

alexey-milovidov committed:
Merge commit '3fa1ed2d9d9cc4f472cd43e2e4eaabac12ec36db' into HEAD

tuanpach committed:
Fix sleep_time backoff overflow and recompute it on recovery

Recompute sleep_time at the single place it is consumed (before
rescheduling), so the Distributed async insert backoff grows after
errors and shrinks again as error_count decays after successful sends.

Extract the delay computation into a pure, unit-tested calculateSleepTime
and make it saturate before multiplying: default_sleep_time * 2^error_count
is an unsigned 64-bit multiply that could wrap to a tiny value (exactly 0
for the default 100ms at error_count 62), defeating the max_sleep_time cap.

Add a gtest covering growth, saturation, decay, the overflow regression,
and the zero base delay edge case.

tuanpach committed:
Merge remote-tracking branch 'origin/master' into fix-descrease-insert-sleep-time-for-distributed-engine

tuanpach committed:
Decay error_count on idle polls and fix clang-tidy init warning

Move the time-gated error_count decay out of the success-only branch to
after the loop, guarded by a had_error flag. This lets the backoff relax
after a shard recovers even when the queue is idle (no files to send),
while still growing the backoff during an ongoing outage. Previously an
idle run exited at !hasPendingFiles() before the decay, so sleep_time
stayed pinned at background_insert_max_sleep_time_ms indefinitely.

Also initialize error_count in updateSleepTime to satisfy
cppcoreguidelines-init-variables (clang-tidy build).

tuanpach committed:
Do not decay error_count while distributed sends are stopped

The idle-poll decay also fired on the monitor_blocker.isCancelled() path,
so SYSTEM STOP DISTRIBUTED SENDS would halve error_count on every blocked
poll even though no shard was retried. After a long stop, SYSTEM START
could resume at the base delay and hammer a still-broken shard.

Guard the decay with !monitor_blocker.isCancelled() so only non-failing,
unblocked runs relax the backoff; a manual stop now preserves it.

alexey-milovidov reviewed (approved):
Looks good.
